kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActiveAddFilesIterator.java

```diff
@@ -0,0 +1,249 @@
+     * dependency between them.
+     *
+     * Ensures that
+     * - `next` is non-empty if there is a next result
+     * - `next` is empty if there is no next result
+     */
+    private void prepareNext() {
+        if (next.isPresent()) return; // already have a next result
+        if (!iter.hasNext()) return; // no next result, and no batches to read
+
+        final Tuple2<FileDataReadResult, Boolean> _next = iter.next();
+        final FileDataReadResult fileDataReadResult = _next._1;
+        final boolean isFromCheckpoint = _next._2;
+        final ColumnarBatch addRemoveColumnarBatch = fileDataReadResult.getData();
+
+        assert(addRemoveColumnarBatch.getSchema().equals(LogReplay.ADD_REMOVE_READ_SCHEMA));
+
+        // Step 1: Update `tombstonesFromJson` with all the RemoveFiles in this columnar batch, if
+        //         and only if this batch is not from a checkpoint.
+        //
+        //         There's no reason to put a RemoveFile from a checkpoint into `tombstonesFromJson`
+        //         since, when we generate a checkpoint, any corresponding AddFile would have
+        //         been excluded already
+        if (!isFromCheckpoint) {
+            final ColumnVector removesVector = addRemoveColumnarBatch.getColumnVector(1);
+            for (int rowId = 0; rowId < removesVector.getSize(); rowId++) {
+                if (removesVector.isNullAt(rowId)) continue;
+
+                // Note: this row doesn't represent the complete RemoveFile schema. It only contains
+                //       the fields we need for this replay.
+
+                final Row row = removesVector.getStruct(rowId);
+                final String path = requireNonNull(row, 0, "path").getString(0);
+                final URI pathAsUri = pathToUri(path);
+                final Optional<String> dvId = Optional.ofNullable(
+                    DeletionVectorDescriptor.fromRow(row.getStruct(1))
+                ).map(DeletionVectorDescriptor::getUniqueId);
+                final UniqueFileActionTuple key = new UniqueFileActionTuple(pathAsUri, dvId);
+                tombstonesFromJson.add(key);
+            }
+        }
+
+        // Step 2: Iterate over all the AddFiles in this columnar batch in order to build up the
+        //         selection vector. We unselect an AddFile when it was removed by a RemoveFile
+        final ColumnVector addsVector = addRemoveColumnarBatch.getColumnVector(0);
+        boolean[] selectionVector = new boolean[addsVector.getSize()];
+        boolean atLeastOneUnselected = false;
+
+        for (int rowId = 0; rowId < addsVector.getSize(); rowId++) {
+            if (addsVector.isNullAt(rowId)) {
+                atLeastOneUnselected = true;
+                continue; // selectionVector will be `false` at rowId by default
+            }
+
+            final Row row = addsVector.getStruct(rowId);
+            final String path = AddFile.getPathFromRow(row);
```

Comment:
According to the PROTOCOL, the path in `AddFile` can be relative or absolute. Do we need to make the paths absolute before comparing? Is there case where the `AddFile` contains relative path, but the absolute path in `RemoveFile`?

Reply:
https://github.com/delta-io/delta/issues/1969